src/agent/coverage/src/block/linux.rs

```diff
@@ -0,0 +1,512 @@
+    }
+
+    Ok(blocks)
+}
+
+pub fn find_symbol_blocks(section: ElfSection, sym: ElfSymbol) -> Result<Vec<Block>> {
+    // Slice symbol's data within section.
+    let lo = (sym.address() - section.address()) as usize;
+    let hi = lo + (sym.size() as usize);
+    let data = section.data()?;
+    let data = &data[lo..hi];
+
+    // Find basic block leaders, as VAs relative to the section load addr.
+    let leaders = find_function_leaders(data, sym);
+
+    let mut decoder = Decoder::new(64, data, DecoderOptions::NONE);
+
+    // Enable using `Decoder::ip()` to get module-relative offsets.
+    decoder.set_ip(sym.address());
+
+    let mut blocks = vec![];
+
+    let mut offset = 0;
+    let mut size = 0;
+
+    let mut inst = Instruction::default();
+    while decoder.can_decode() {
+        let current = decoder.ip() as u64;
+
+        decoder.decode_out(&mut inst);
+
+        let is_leader = leaders.contains(&current);
+
+        // May not point to an actual instruction in the function span.
+        let next = decoder.ip() as u64;
+
+        // If the next instruction is a leader, or the end of the function, then
+        // we have reached the end of the block.
+        let end = !decoder.can_decode() || leaders.contains(&next);
+
+        size += inst.len() as u64;
+
+        if is_leader {
+            offset = current;
+        }
+
+        // Note: for a 1-instruction block, `is_leader && end`.
+        if end {
+            blocks.push(Block { offset, size });
+
+            // Reset WIP block.
+            offset = 0;
+            size = 0;
+        }
+    }
+
+    Ok(blocks)
+}
+
+/// Compute the basic block leaders of a function as virtual addresses.
+fn find_function_leaders(data: &[u8], sym: ElfSymbol) -> BTreeSet<u64> {
+    let mut decoder = Decoder::new(64, data, DecoderOptions::NONE);
+    decoder.set_ip(sym.address());
+
+    // Contains leaders with VAs, assuming section load address.
+    let mut leaders = BTreeSet::new();
+
+    // Function entry is a leader.
+    leaders.insert(sym.address());
+
+    let mut inst = Instruction::default();
+    while decoder.can_decode() {
+        decoder.decode_out(&mut inst);
+
+        if let Some(target) = branch_target(&inst) {
+            // The branch target is a leader.
+            leaders.insert(target);
+
+            // The next instruction is a leader, if it exists.
+            if decoder.can_decode() {
+                // We decoded the current instruction, so the decoder offset is
+                // set to the next instruction.
+                let next = decoder.ip() as u64;
+                leaders.insert(next);
+            }
+        }
+    }
+
+    leaders
+}
+
+fn branch_target(inst: &iced_x86::Instruction) -> Option<u64> {
+    use iced_x86::FlowControl;
+
+    match inst.flow_control() {
+        FlowControl::ConditionalBranch | FlowControl::UnconditionalBranch => {
+            Some(inst.near_branch_target())
+        }
+        _ => None,
```

Comment:
I do have a slight preference to specify the cases explicitly here (as in the Windows equivalent code) because those cases do feel like a TODO (low priority) instead of a don't care.